### PR TITLE
Add `ScriptNum` type

### PR DESCRIPTION
**Pull request by tcharding**
Draft for feedback please.  Also there is one minor FIXME in there that needs resolving.

Only the last two patches.

Does not need to be reviewed or considered for merge until after 0.30.0 release.

This was a very educational PR to work on - I think I now get how integers work in Bitcoin Script :)

Feedback on whether my testing efforts add meaningful value also please (primarily the kani tests).

### CODE REVIEW

File: bitcoin/src/blockdata/script/num.rs [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+/// - First encode the absolute value of `x` as a little-endian array of bytes.
+/// - If the most significant byte is >= 0x80 and the value is positive, add a
+///   new zero-byte to make the significant byte < 0x80 again.
+/// - If the most significant byte is >= 0x80 and the value is negative, add a
+///   new 0x80 byte that will be popped off when decoding.
+/// - If the most significant byte is < 0x80 and the value is negative, add a
+///   now 0x80 byte, since it will be subtracted and interpreted as a negative
+///   when decoding.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ScriptNum(pub i64);
```
Comment (Kixunil) [inline bitcoin/src/blockdata/script/num.rs:61]: Why is this `pub`? The entire point of the type was to avoid invalid values!

**Review by Kixunil — COMMENTED** [review]
I didn't sleep well tonight so I might be missing something but this doesn't match the design I was thinking about.

File: bitcoin/src/blockdata/script/num.rs [hunk 2]
```diff
@@ -0,0 +1,570 @@
+//! Provides the [`ScriptNum`] type which is equivalent to `CScriptNum` from Bitcoin Core.
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply bitcoin/src/blockdata/script/num.rs:1]: Yes, and yes. Its purpose is to encode numbers in the Script encoding, which is a weird signed-magnitude thing.

With the philosophical difference that in Core you can construct `ScriptNum`s directly from values that are out of range of the numeric opcodes, while (I think) we don't intend to provide such a constructor. Or if we do, we should give it a scary name.

File: bitcoin/src/blockdata/script/num.rs (hunk 1 above)
Comment (apoelstra) [reply bitcoin/src/blockdata/script/num.rs:61]: Strictly speaking there are no "invalid values" for a ScriptNum, only values which the numeric opcodes will reject. (Amusingly, these opcodes will *output* such values.)

But I agree, we should hide the inner type and add a `new_from_i64` method or something with a big doccomment warning that if you provide out-of-range values you're probably going to crash your script, and that you probably don't want to use this constructor.

**Comment by tcharding** [comment]
The discussion is fragmented, its on https://github.com/rust-bitcoin/rust-bitcoin/issues/1530 as well.

I think I'll try to put together a PR that does #1530, #1547, and this all together so we have a single place to discuss it.

**Pull request closed without merging**